authorize/evaluator/policy_evaluator.go

```diff
@@ -145,41 +156,49 @@ func NewPolicyEvaluator(
 		}
 	}
 
-	// for each script, create a rego and prepare a query.
-	for i := range e.queries {
-		log.Debug(ctx).
-			Str("script", e.queries[i].script).
-			Str("from", configPolicy.From).
-			Interface("to", configPolicy.To).
-			Msg("authorize: rego script for policy evaluation")
-
-		r := rego.New(
-			rego.Store(store),
-			rego.Module("pomerium.policy", e.queries[i].script),
-			rego.Query("result = data.pomerium.policy"),
-			rego.EnablePrintStatements(true),
-			getGoogleCloudServerlessHeadersRegoOption,
-			store.GetDataBrokerRecordOption(),
-		)
-
-		q, err := r.PrepareForEval(ctx)
-		// if no package is in the src, add it
-		if err != nil && strings.Contains(err.Error(), "package expected") {
-			r := rego.New(
-				rego.Store(store),
-				rego.Module("pomerium.policy", "package pomerium.policy\n\n"+e.queries[i].script),
-				rego.Query("result = data.pomerium.policy"),
-				rego.EnablePrintStatements(true),
-				getGoogleCloudServerlessHeadersRegoOption,
-				store.GetDataBrokerRecordOption(),
-			)
-			q, err = r.PrepareForEval(ctx)
-		}
-		if err != nil {
-			return nil, err
+	// for each script, create a rego object and prepare a query.
+	rttrace.WithRegion(ctx, "Compile Rego", func() {
+		numCached := 0
+		for i := range e.queries {
+			var cached bool
+			e.queries[i].PreparedEvalQuery, cached, err = cache.LookupOrBuild(&e.queries[i], func() (rego.PreparedEvalQuery, error) {
+				r := rego.New(
+					rego.Store(store),
```

Comment:
Looking at the code there's a potential issue here involving the `store` variable. If it were changed I don't think we'd detect the change and end up using a prepared query with the wrong store.

That said, it looks like we never actually change the `store` so this bug won't happen. But it seems like a footgun.

Reply:
Good catch, I will add cache invalidation here if the store changes.